deker_server_adapters/factory.py

```diff
@@ -116,28 +123,83 @@ def get_collection_adapter(
+        :param client: Httpx Client
+        :param in_cluster: If we are in cluster
         """
-        response = None
-        nodes = [*ctx.uri.servers]
-        while nodes and (response is None or response.status_code != STATUS_OK):
-            node = nodes.pop()
 
+        def check_response(response: Optional[Response], client: HttpxClient) -> None:
+            if response is None or response.status_code != STATUS_OK:
+                client.close()
+                raise DekerServerError(
+                    response,
+                    "Healthcheck failed. Deker client will be closed.",
+                )
+
+        url = "v1/ping"
+
+        # If we do healthcheck in cluster
+        nodes = [*ctx.uri.servers] if in_cluster else [ctx.uri.raw_url]
+        response = make_request(url=url, nodes=nodes, client=client)
+        check_response(response=response, client=client)
+
+        if in_cluster:
             try:
-                response = self.httpx_client.get(f"{node}/v1/ping")
-            except Exception:
-                self.logger.error(f"Coudn't get response from {node}")  # noqa
-                traceback.print_exc()
-                continue
-        if response is None or response.status_code != STATUS_OK:
-            self.httpx_client.close()
-            raise DekerServerError(
-                response,
-                "Healthcheck failed. Deker client will be closed.",
-            )
-
-        # set hash_ring based on list from the server
-        ctx.extra["hash_ring"] = HashRing(response.json()["servers"])
+                config = response.json()  # type: ignore[union-attr]
+                return config
+            except JSONDecodeError:
+                raise DekerClusterError(response, "Server responded with empty config")
+
+    def __set_cluster_config(self, cluster_config: Dict, ctx: CTX) -> None:
+        """Set cluster config in the CTX.
+
+        :param cluster_config: Custer config json from server
+        :param ctx: App cotext (Deker CTX)
+        """
+        # IDs used in hash ring
+        ids = []
+        # Mapping from ID to host
+        id_to_host_mapping = defaultdict(list)
+        leader_node = None
+        nodes = []
+
+        # Fill Ids and Mappings
+        for node in cluster_config["current_nodes"]:
+            url = f"{node['protocol']}://{node['host']}:{node['port']}"
+            nodes.append(url)
+            ids.append(node["id"])
+            id_to_host_mapping[node["id"]].append(url)
+            if node["id"] == cluster_config["leader_id"]:
+                leader_node = url
+
+        if leader_node is None:
+            raise DekerServerError(None, f"Leader node cannot be setted {cluster_config=}")
+
+        # Set variables in context
+        ctx.extra["leader_node"] = Uri.create(leader_node)
+        ctx.extra["nodes_mapping"] = id_to_host_mapping
+        ctx.extra["hash_ring"] = HashRing(ids)
+        ctx.extra["nodes"] = nodes
```

Comment:
Perhaps it is time to refactor `extra` from dict into a class object with preset attributes? It shall avoid keys/attributes naming mash up.

Reply:
I got you, but cannot agree. Using dict is very convenient as it allows us to pass extra option without changing signature in the deker package. As I remember we don't want to have mentions of deker server or server adapters inside embedded version